pkg/syncer/fullsync.go

```diff
@@ -272,13 +275,67 @@ func CsiFullSync(ctx context.Context, metadataSyncer *metadataSyncInformer, vc s
 	go fullSyncDeleteVolumes(ctx, volToBeDeleted, metadataSyncer, &wg, migrationFeatureStateForFullSync, volManager, vc)
 	wg.Wait()
 
+	// Sync VolumeInfo CRs
+	if isMultiVCenterFssEnabled && len(metadataSyncer.configInfo.Cfg.VirtualCenter) > 1 {
+		volumeInfoCRFullSync(ctx, k8sPVMap, vc)
+	}
+
 	cleanupCnsMaps(k8sPVMap)
 	log.Debugf("FullSync: cnsDeletionMap at end of cycle: %v", cnsDeletionMap)
 	log.Debugf("FullSync: cnsCreationMap at end of cycle: %v", cnsCreationMap)
 	log.Infof("FullSync: end")
 	return nil
 }
 
+// volumeInfoCRFullSync creates VolumeInfo CR if it does not already exist for a volume.
+// It also deletes VolumeInfo CR if its corresponding PV does not exist.
+func volumeInfoCRFullSync(ctx context.Context, k8sPvs map[string]string, vc string) {
+	log := logger.GetLogger(ctx)
+
+	for volumeID := range k8sPvs {
+		crExists, err := volumeInfoService.VolumeInfoCrExistsForVolume(ctx, volumeID)
+		if err != nil {
+			log.Errorf("FullSync: failed to find VolumeInfo CR for volume %s."+
+				"Error: %+v", volumeID, err)
+			continue
+		}
+
+		// Create VolumeInfo CR if not found.
+		if !crExists {
+			err := volumeInfoService.CreateVolumeInfo(ctx, volumeID, vc)
+			if err != nil {
+				log.Errorf("FullSync: failed to create VolumeInfo CR for volume %s."+
+					"Error: %+v", volumeID, err)
+				continue
+			}
+		}
+	}
+
+	volumeInfoCRList := volumeInfoService.ListAllVolumeInfos()
+	for _, volumeInfo := range volumeInfoCRList {
+		cnsvolumeinfo := &cnsvolumeinfov1alpha1.CNSVolumeInfo{}
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(volumeInfo.(*unstructured.Unstructured).Object,
+			&cnsvolumeinfo)
+		if err != nil {
+			log.Errorf("FullSync: failed to parse cnsvolumeinfo object: %v, err: %v", cnsvolumeinfo, err)
+			continue
+		}
+
+		if cnsvolumeinfo.Spec.VCenterServer == vc {
```

Comment:
is it possible to have same volumeID across multiple VCs ? if not, we may not need this check for vc match

Reply:
No we cannot have same volumeID across VCs.

But we still need this check because we will get all CRs in the cluster which can be from any of the VCs in the deployment. However, we're only concerned with CRs int he given VC.